### PR TITLE
Decouple Phragmen from Staking. 

### PR DESCRIPTION
**Pull request by kianenigma**
For the sake of it being used in two modules now (Staking and new council) I want to both clean and optimize the phragmen code to some extend. 

So what happened is that:

- I partially moved the main chunk of it to sr-primitives in #3364 .
- I also started some improvements to it #3456.

At this point, I don't want to take #3356 any further because I want to rebase everything into a phragmen implementation which is _100% decoupled from staking module_. I assumed #3364 would get merged easily but I don't see that happening anytime soon because @gavofyork is offline this week. Hence, this PR is to override this. Also renames some terms and types to make it independent of Staking.

As soon as it is done, both of the mentioned PRs above should be rebased with these changes. From that point onward #3364 would stay unchanged and just needs a merge and #3456 can proceed without the problem of a merge-hell later on

An enhancement would be to move the benchmarks and tests of phragmen out of staking but I will not do that in this PR to minimize the changes.

### CODE REVIEW

**Comment by bkchr** [comment]
Why not move it into an own crate? This is an algorithm and sounds like something that can happily live in its own space. In the future, I would even expect that this crate live in its own repository like Grandpa.

**Comment by kianenigma** [comment]
I am totally in for that but once the tests + benchmarked + reference impl with f64 is also provided. If you agree with my roadmap above and why I really want this decoupling to push my other two PRs forward, for now, we can start reviewing this (it is almost insubstantial in terms of logic change) and migrate to a crate soonish.

**Comment by kianenigma** [comment]
Well, it is possible but moving to its own crate needs a bit more work since we are using a few traits/type from substrate primitives. Not sure how valuable will the migration be if we still depend on substrate?

**Comment by bkchr** [comment]
In a first step I meant just its own crate in this repository.

File: srml/staking/src/lib.rs [hunk 1]
```diff
@@ -1339,11 +1317,24 @@ impl<T: Trait> Module<T> {
...
-				if e.total < slot_stake {
-					slot_stake = e.total;
+			for (c, s) in supports.into_iter() {
+				// build `struct exposure` from `support`
+				let exposure = Exposure {
+					own: to_balance(s.own),
+					// This might reasonably saturate and we cannot do much about it. The sum of
+					// someone's stake might exceed the balance type if they have the maximum amount
+					// of balance and receive some support. This is super unlikely to happen, yet
+					// we simulate it it in some tests.
```
Comment (kianenigma) [inline srml/staking/src/lib.rs]:
```suggestion
					// we simulate it in some tests.
```

File: core/phragmen/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,19 @@
+[package]
+name = "phragmen"
```
Comment (bkchr) [inline core/phragmen/Cargo.toml]:
```suggestion
name = "substrate-phragmen"
```

File: srml/staking/src/benches.rs [hunk 3]
```diff
@@ -71,67 +72,55 @@ fn do_phragmen(
...
-					a.into_iter().map(|(acc, r)| (
-						acc.clone(),
-						r,
-						to_balance(ratio_of(Staking::slashable_balance_of(&n), r)),
-					))
-					.collect::<Vec<Assignment<Test>>>()
-				)).collect::<Vec<(AccountId, Balance, Vec<Assignment<Test>>)>>();
-
-				let mut exposures = <ExpoMap<Test>>::new();
+					<<Test as Trait>::CurrencyToVote as Convert<Balance, AccountId>>::convert(b) as ExtendedBalance;
```
Comment (bkchr) [inline srml/staking/src/benches.rs]:
```suggestion
				<<Test as Trait>::CurrencyToVote as Convert<Balance, AccountId>>::convert(b) as ExtendedBalance;
```

File: core/phragmen/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,726 @@
...
+	candidate_count: usize,
+	minimum_candidate_count: usize,
+	initial_candidates: Vec<AccountId>,
+	initial_voters: Vec<(AccountId, Vec<AccountId>)>,
+	stake_of: FS,
+	self_vote: bool,
+) -> Option<PhragmenResult<AccountId>> where
+	AccountId: Default + Ord + Member,
+	Balance: Default + Copy + SimpleArithmetic,
+	for <'r> FS: Fn(&'r AccountId) -> Balance,
```
Comment (bkchr) [inline core/phragmen/src/lib.rs]:
```suggestion
	for<'r> FS: Fn(&'r AccountId) -> Balance,
```

File: core/phragmen/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,726 @@
...
+	let mut voters: Vec<Voter<AccountId>> = Vec::with_capacity(num_voters);
+
+	// collect candidates. self vote or filter might apply
+	let mut candidates = if self_vote {
+		// self vote. filter.
+		initial_candidates.into_iter().map(|who| {
+			let stake = stake_of(&who);
+			Candidate { who, approval_stake: to_votes(stake), ..Default::default() }
+		})
+		.filter_map(|c| {
```
Comment (bkchr) [inline core/phragmen/src/lib.rs]: You want to have a: `filter(|c| !c.approval_stake.is_zero())`

File: core/phragmen/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,726 @@
...
+
+		for (voter, assignment) in assignments.iter_mut() {
+			let voter_budget = stake_of(&voter);
+
+			let diff = do_equalize::<_, _, C>(
+				voter,
+				voter_budget,
+				assignment,
+				supports,
+				tolerance
```
Comment (bkchr) [inline core/phragmen/src/lib.rs]:
```suggestion
				tolerance,
```

**Pull request merged**